File: android/core/model/src/main/kotlin/mv/muraka/core/model/Sighting.kt
package mv.muraka.core.model

import java.time.Instant

/** A coordinate in WGS84, the only spatial type the client needs. */
data class Position(val lat: Double, val lon: Double) {
    val isValid: Boolean get() = lat in -90.0..90.0 && lon in -180.0..180.0
}

/**
 * A sighting as the **server** holds it.
 *
 * There is deliberately no `synced` field and no client-computed status. This type only
 * ever comes from a server response; anything the client knows about work it has not
 * delivered yet lives in the outbox instead, and the two are combined for display by
 * `SightingDisplayStatus`. See D21 in `docs/08` for why that separation is the whole
 * point rather than an implementation detail.
 */
data class Sighting(
    val id: String,
    val contributorId: String,
    val contributorName: String? = null,
    val siteId: String? = null,
    val siteName: String? = null,
    val position: Position,
    val locationSource: LocationSource,
    val locationAccuracyM: Double? = null,
    val depthM: Double? = null,
    val capturedAt: Instant,
    val note: String? = null,
    val selfAssessedCondition: Condition? = null,
    val status: SightingStatus,
    /** When the server received it, as distinct from when it was captured. */
    val createdAt: Instant,
    val photoCount: Int = 0,
    /**
     * Effective condition: the expert's label when one exists, otherwise the model's.
     * Absent until analysis completes. Never render this without [verified] beside it —
     * a model label presented as fact is the NFR13 failure.
     */
    val condition: Condition? = null,
    /** Worst bleached extent across the sighting's photos, 0–1. */
    val severity: Double? = null,
    val confidence: Double? = null,
    /** True only when an expert confirmed or corrected. */
    val verified: Boolean = false,
)

/** One photograph belonging to a sighting, with the model's reading of it if it exists. */
data class Photo(
    val id: String,
    val sightingId: String,
    /** Relative path to the bytes. Requires the bearer token — it is not a public URL. */
    val url: String,
    val width: Int,
    val height: Int,
    val bytes: Int,
    val createdAt: Instant,
    /** Absent until classification finishes. Absent is not an error. */
    val prediction: Prediction? = null,
)

/** One cell of the inference grid. */
data class Patch(
    val row: Int,
    val col: Int,
    val label: Condition,
    val confidence: Double,
)

/**
 * What the model made of one photograph.
 *
 * [severity] is the number to lead with, not [label]: "62% bleached" tells a contributor
 * something "bleached" does not. [modelVersion] is provenance and must be shown —
 * `fake-0.0.0` means no trained model is loaded yet (D19).
 */
data class Prediction(
    val id: String,
    val photoId: String,
    val modelVersion: String,
    val label: Condition,
    val confidence: Double,
    val severity: Double,
    val patchGrid: Int,
    val patches: List<Patch>,
    val inferenceMs: Int? = null,
    val createdAt: Instant,
)

/** An expert's decision on a sighting. */
data class Verification(
    val id: String,
    val sightingId: String,
    val verifierId: String,
    val verifierName: String? = null,
    val decision: VerificationDecision,
    val label: Condition? = null,
    val rejectReason: RejectReason? = null,
    val comment: String? = null,
    val createdAt: Instant,
)

/** Everything `GET /v1/sightings/{id}` returns. */
data class SightingDetail(
    val sighting: Sighting,
    val photos: List<Photo>,
    val verifications: List<Verification>,
)

File: ios/Muraka/Domain/Model/Sighting.swift
import Foundation

/// A coordinate in WGS84.
struct Position: Equatable, Sendable {
    let lat: Double
    let lon: Double

    /// The server rejects anything outside these bounds with a terminal `422`.
    static let latitudeRange = -90.0 ... 90.0
    static let longitudeRange = -180.0 ... 180.0

    var isValid: Bool {
        Self.latitudeRange.contains(lat) && Self.longitudeRange.contains(lon)
    }
}

/// A sighting as the **server** holds it.
///
/// There is deliberately no `synced` field and no client-computed status. This type only
/// ever comes from a server response; what the client knows about work it has not
/// delivered lives in the outbox, and the two are combined only for display. See D21.
struct Sighting: Identifiable, Equatable, Sendable {
    let id: String
    let contributorID: String
    var contributorName: String?
    var siteName: String?
    let position: Position
    let locationSource: LocationSource
    var locationAccuracyM: Double?
    var depthM: Double?
    let capturedAt: Date
    var note: String?
    var selfAssessedCondition: Condition?
    let status: SightingStatus
    /// When the server received it, as distinct from when it was captured.
    let createdAt: Date
    var photoCount: Int = 0
    /// Expert label where one exists, otherwise the model's. Never render without
    /// ``verified`` beside it — that is the NFR13 failure.
    var condition: Condition?
    /// Worst bleached extent across the sighting's photographs, 0–1.
    var severity: Double?
    var confidence: Double?
    /// True only when an expert confirmed or corrected.
    var verified: Bool = false
}

/// One cell of the inference grid.
struct Patch: Equatable, Sendable {
    let row: Int
    let col: Int
    let label: Condition
    let confidence: Double
}

/// What the model made of one photograph.
///
/// ``severity`` is the number to lead with, not ``label``: "62% bleached" tells a
/// contributor something "bleached" does not. ``modelVersion`` is provenance and must be
/// shown — `fake-0.0.0` means no trained model is loaded yet (D19).
struct Prediction: Equatable, Sendable {
    let id: String
    let photoID: String
    let modelVersion: String
    let label: Condition
    let confidence: Double
    let severity: Double
    let patchGrid: Int
    let patches: [Patch]
    var inferenceMs: Int?
    let createdAt: Date
}

/// One photograph, with the model's reading of it if it exists.
struct Photo: Identifiable, Equatable, Sendable {
    let id: String
    let sightingID: String
    /// Relative path to the bytes. Requires the bearer token — not a public URL.
    let url: String
    let width: Int
    let height: Int
    let bytes: Int
    let createdAt: Date
    /// Absent until classification finishes. Absent is not an error.
    var prediction: Prediction?
}

/// An expert's decision on a sighting.
struct Verification: Identifiable, Equatable, Sendable {
    let id: String
    let sightingID: String
    var verifierName: String?
    let decision: VerificationDecision
    var label: Condition?
    var rejectReason: RejectReason?
    var comment: String?
    let createdAt: Date
}

/// Everything `GET /v1/sightings/{id}` returns.
struct SightingDetail: Equatable, Sendable {
    let sighting: Sighting
    var photos: [Photo] = []
    var verifications: [Verification] = []
}
